src/rust/ide/src/controller/upload.rs

```
@@ -0,0 +1,625 @@
+        }
+    }
+
+    fn metadata_of_new_node<DP>(file:&FileToUpload<DP>, position:Position) -> NodeMetadata {
+        let uploading_metadata = UploadingFile {
+            name           : file.name.clone(),
+            remote_name    : None,
+            size           : file.size,
+            bytes_uploaded : 0,
+            error          : None,
+        };
+        NodeMetadata {
+            position        : Some(position),
+            intended_method : None,
+            uploading_file  : Some(uploading_metadata)
+        }
+    }
+
+    async fn upload_file
+    (&self, node:ast::Id, file:FileToUpload<impl DataProvider>) -> FallibleResult {
+        self.ensure_data_directory_exists().await?;
+        let remote_name = self.establish_remote_file_name(&file.name).await?;
+        self.update_metadata(node, |md| md.remote_name = Some(remote_name.clone()));
+        self.update_expression(node,Self::uploading_node_expression(&remote_name))?;
+        let remote_path     = self.data_path().append_im(&remote_name);
+        let bin_connection  = self.project.binary_rpc();
+        let json_connection = self.project.json_rpc();
+        let mut process     = FileUploadProcess::new
+            (&self.logger,file,bin_connection,json_connection,remote_path);
+
+        while !process.upload_chunk().await? {
+            self.update_metadata(node, |md| md.bytes_uploaded = process.bytes_uploaded);
+        }
+        self.update_expression(node,Self::uploaded_node_expression(&remote_name))?;
+        if let Err(err) = self.graph.module.with_node_metadata(node, Box::new(|md| md.uploading_file = None)) {
+            warning!(self.logger, "Cannot remove uploading metadata: {err}");
+        }
+        Ok(())
+    }
+
+    fn update_metadata(&self, node:ast::Id, f:impl FnOnce(&mut UploadingFile)) {
+        //TODO[ao] see the TODO comment in update_expression.
+        let _tr = self.undo_redo_repository().open_ignored_transaction("Upload Metadata Update");
```

`open_ignored_transaction` returns a `Result`. Is it fine to ignore the error case?

---

Yes. The open_ignored_transaction returns Err if there is already opened transaction: but this is normal.